Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright (C) 2022 Igalia S.L.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE COMPUTER, INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE COMPUTER, INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "NotificationService.h"
+
+#include "WebNotification.h"
+#include <gio/gio.h>
+#include <glib/gi18n-lib.h>
+#include <mutex>
+#include <wtf/FastMalloc.h>
+#include <wtf/Seconds.h>
+#include <wtf/UUID.h>
+#include <wtf/glib/GUniquePtr.h>
+#include <wtf/glib/Sandbox.h>
+#include <wtf/text/CString.h>
+
+#if PLATFORM(GTK)
+#include <WebCore/GtkVersioning.h>
+#endif
+
+namespace WebKit {
+
+static const Seconds s_dbusCallTimeout = 20_ms;
+
+NotificationService& NotificationService::singleton()
+{
+    static std::once_flag onceFlag;
+    static LazyNeverDestroyed<NotificationService> service;
+
+    std::call_once(onceFlag, [] {
+        service.construct();
+    });
```

Comment:
This class is not threadsafe, so you should just store a bool rather than using `std::call_once`. The point of `std::call_once` is to use a mutex to guarantee it gets initialized once regardless of thread.

Reply:
This is a common pattern in all singleton using LazyNeverDestroyed, I think it's harmless even if the class it not thread safe.